src/aws_encryption_sdk/internal/utils/streams.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Helper stream utility objects for AWS Encryption SDK."""
+import io
+
+from aws_encryption_sdk.exceptions import ActionNotAllowedError
+
+
+class PassThroughStream(object):
+    """Provides a pass-through interface on top of a stream object.
+
+    :param source_stream: File-like object
+    """
+
+    def __init__(self, source_stream):
+        """Prepares the passthroughs."""
+        self._source_stream = source_stream
+        self._duplicate_api()
+
+    def _duplicate_api(self):
```

Comment:
Good catch on the attributes/methods. Fixed that.

The problem with just subclassing `io.BaseIO` is that these are really meant to front file-like objects, not just streams (fixed that terminology in the comments too). File-like objects (of which streams are a subset) have a consistent API but it is also quite large, so I would rather avoid manually duplicating it.

Effectively what this is doing is duplicating the public API of `_source_stream` onto the instance of `PassThroughStream` in the form of pointers to the attributes on `_source_stream`. This makes it so that any public interactions with the instance of `PassThroughStream` behave exactly as they would if they were made to `_source_stream` directly. Behavior of calls inside `_source_stream` are unaffected, as they are still happening within the context of that (unmodified) object.

Aside from the obvious overrides, the only place where this should introduce any unexpected behavior is if someone is making calls to private-hinted attributes on `_source_stream`, and I think that is a safe place to draw the line, and it matches with the [convention laid out in PEP8](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces).

Reply:
Per offline discussion: replacing `PassThroughStream` with `ProxyObject` from [wrapt](http://wrapt.readthedocs.io/en/latest/).

Reply:
Yeah, sorry for my confusion on the API duplication, I did not remember that interaction correctly. wrapt is a good find, that turned out as a nice clean drop-in.